// akinator/index.js
const EventEmitter = require('events');
const request = require('snekfetch');

const Endpoint = 'http://api-usa4.akinator.com/ws';

class Akinator extends EventEmitter {
  constructor() {
    super();
    this.identification = {};
  }

  start() {
    return this.req('/new_session', { query: { partner: 0 } })
      .then(({ identification, step_information: q }) => {
        this.identification = identification;
        this.emit('question', Object.assign(q, {
          answer: (n) => this.answer(n, q.step),
        }));
      });
  }

  answer(num, step) {
    return this.req('/answer', { query: {
      answer: num,
      step,
    } })
      .then((q) => {
        const next = () => {
          this.emit('question', Object.assign(q, {
            answer: (n) => this.answer(n, q.step),
          }));
        };
        if (q.progression > 90) {
          this.getList(q.step).then(({ elements }) => {
            this.emit('guess', Object.assign(elements[0].element, {
              guess: (correct) => correct ? this.emit('end', true) : next(),
            }));
          });
        } else {
          return next();
        }
      });
  }

  getList(step) {
    return this.req('/list', { query: {
      step,
    } });
  }

  req(path, options) {
    return request.get(`${Endpoint}${path}`)
      .query(Object.assign({
        session: this.identification.session,
        signature: this.identification.signature,
      }, options.query || {}))
      .then((r) => r.body.parameters);
  }
}

module.exports = Akinator;
